clang/lib/AST/ItaniumMangle.cpp

```diff
@@ -2586,8 +2561,9 @@ bool CXXNameMangler::mangleUnresolvedTypeOrSimpleId(QualType Ty,
   }
 
   case Type::InjectedClassName:
-    mangleSourceNameWithAbiTags(
-        cast<InjectedClassNameType>(Ty)->getDecl());
+    mangleSourceNameWithAbiTags(cast<InjectedClassNameType>(Ty)
+                                    ->getOriginalDecl()
+                                    ->getDefinitionOrSelf());
```

Comment:
Is this `getDefinitionOrSelf()` call necessary here? Shouldn't `InjectedClassNameType` always refer to the definition where the class name is injected to?

Reply:
Not always.

I mean you could certainly attempt to define an `InjectedClassNameType` whose declaration necessarily is the definition.

This would cause many annoyances, most of them surmountable, as you wouldn't be able to form a type to an arbitrary declaration (which we do offhandedly in error recovery and printing diagnostics), and the type would play by different rules than other tag types.

Though I think at the end, you would come to the problem of demoted definitions, which come up for example when merging definitions from the GMF of different modules. You would have to throw away the ability to represent an InjectedClassNameType pointing to a demoted definition.

Reply:
I've caught an example:
```cpp
template <typename>
struct Tpl;

template <typename>
struct Tpl {
  Tpl(const Tpl&) {}
};
```
The injected name from the copy ctor prototype refers to the forward-declaration from the first two lines. Looks a little strange but Ok.

Reply:
Are you sure it's not just its canonical type which refers to the first two lines?

Reply:
Heh, interesting. `getAs` function actually returns the canonical type.

Reply:
getAs only desugars. This could happen with a TemplateSpecializationType which its canonical type is an InjectedClassNameType, since in that case desugar just returns the canonical type, but that is not the case for that constructor, since its using an InjectedClassNameType as written in source code.

Reply:
I've understood: there are specializations of `getAs` for the leaf non-sugar types:

https://github.com/llvm/llvm-project/blob/71925a90c8b713d8fccbfae496e10c68f80b932b/clang/include/clang/AST/Type.h#L3138-L3140

If you call `getAs<InjectedClassNameType>()` on a `Type` which is already `InjectedClassNameType`, you get its canonical type.

Reply:
Btw, it is a drawback of templates that you can easily overlook a specialization.

Reply:
This will be fixed here: https://github.com/llvm/llvm-project/pull/155028